Rename FileSystem::FileOpenMode::Write to FileSystem::FileOpenMode::Truncate

szewai opened the pull request:
#### 22004f62ebe45a346a22574601e6b5191cc62cc5
<pre>
Rename FileSystem::FileOpenMode::Write to FileSystem::FileOpenMode::Truncate
<a href="https://bugs.webkit.org/show_bug.cgi?id=251073">https://bugs.webkit.org/show_bug.cgi?id=251073</a>
rdar://104590894

Reviewed by Žan Doberšek.

In existing implementation, file opened with FileSystem::FileOpenMode::Write will be truncated or recreated if it
exists. This behavior is not clearly indicated in the name, and it causes confusion to users. For example, some places
check file size or truncate file after opening it with FileSystem::FileOpenMode::Write mode. They may actually need to
use FileSystem::FileOpenMode::ReadWrite mode. Therefore, this patch renames FileSystem::FileOpenMode::Write to
FileSystem::FileOpenMode::Truncate. Also, this patch makes glib implementation of FileSystem::FileOpenMode::Truncate
mode aligned with POSIX and win implementation, which recreates file when file with same name already exists.

* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/WasmStreamingParser.cpp:
(JSC::Wasm::dumpWasmSource):
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::MappedFileData::mapFileHandle):
(WTF::FileSystemImpl::readOrMakeSalt):
(WTF::FileSystemImpl::overwriteEntireFile):
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::FileSystemImpl::openFile):
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::openFile):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::openFile):
(WTF::FileSystemImpl::MappedFileData::mapFileHandle):
* Source/WebCore/Modules/webdatabase/OriginLock.cpp:
* Source/WebCore/loader/appcache/ApplicationCacheStorage.cpp:
(WebCore::ApplicationCacheStorage::writeDataToUniqueFileInDirectory):
* Source/WebCore/platform/network/BlobRegistryImpl.cpp:
(WebCore::BlobRegistryImpl::writeBlobToFilePath):
* Source/WebCore/platform/network/curl/CookieJarDB.cpp:
(WebCore::CookieJarDB::flagDatabaseCorruption):
* Source/WebCore/platform/network/curl/CurlCacheEntry.cpp:
(WebCore::CurlCacheEntry::saveResponseHeaders):
(WebCore::CurlCacheEntry::openContentFile):
* Source/WebCore/platform/network/curl/CurlCacheManager.cpp:
(WebCore::CurlCacheManager::saveIndex):
* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::store):
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::NetworkDataTaskBlob::download):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
(WebKit::ServiceWorkerDownloadTask::setPendingDownloadLocation):
* Source/WebKit/NetworkProcess/cache/CacheStorageEngine.cpp:
(WebKit::CacheStorage::Engine::writeSizeFile):
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::dumpContentsToFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCurl.cpp:
(WebKit::NetworkCache::IOChannel::IOChannel):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::invokeDidReceiveResponse):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::writeCachesList):
* Source/WebKit/Shared/PersistencyUtils.cpp:
(WebKit::writeToDisk):
* Source/WebKit/Shared/WebMemorySampler.cpp:
(WebKit::WebMemorySampler::initializeSandboxedLogFile):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::writeSandboxDataToCacheFile):
* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::ContentRuleListStore::invalidateContentRuleListVersion):
* Source/WebKit/UIProcess/ios/WKModelView.mm:
(-[WKModelView createFileForModel:]):
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.mm:
(WebKit::ARKitInlinePreviewModelPlayerMac::createFile):
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
(TestWebKitAPI::createTestFile):
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebCore/FileMonitor.cpp:
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/DatabaseTrackerTest.mm:
(TestWebKitAPI::createFileAtPath):

Canonical link: <a href="https://commits.webkit.org/259689@main">https://commits.webkit.org/259689@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/56492e78df52958a8f1f9cf44246eed99d960f97

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105632 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14723 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38507 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114860 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175011 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109534 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16129 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5922 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97899 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114678 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111386 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12264 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95244 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39744 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94128 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26883 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/95365 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7991 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28236 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/93531 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5774 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8491 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30407 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14108 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47784 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/102244 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6696 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10042 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25496 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

zdobersek commented on Source/WTF/wtf/glib/FileSystemGlib.cpp:
```diff
@@ -195,10 +195,18 @@ PlatformFileHandle openFile(const String& path, FileOpenMode mode, FileAccessPer
+                fileExists = false;
+            } 
+        }
+        if (!fileExists)
```
I think `g_file_replace_readwrite()` can be used here for truncation mode, and the whole logic of this function could probably be reworked a bit to accommodate it. Somebody else can swoop in and do it after this change.

mcatanzaro replied:
I took a look and I think the code should just be:

```
    if (mode == FileOpenMode::Read || mode == FileOpenMode::ReadWrite)
        ioStream = adoptGRef(g_file_open_readwrite(file.get(), nullptr, nullptr));
    else {
        ASSERT(mode == FileOpenMode::Truncate);
        ioStream = adoptGRef(g_file_replace_readwrite(file.get(), nullptr, FALSE, permissionFlag, nullptr, nullptr));
    }
```

In patch form:

```
diff --git a/Source/WTF/wtf/glib/FileSystemGlib.cpp b/Source/WTF/wtf/glib/FileSystemGlib.cpp
index 3b522e6bb085..9941909185d1 100644
--- a/Source/WTF/wtf/glib/FileSystemGlib.cpp
+++ b/Source/WTF/wtf/glib/FileSystemGlib.cpp
@@ -193,21 +193,11 @@ PlatformFileHandle openFile(const String& path, FileOpenMode mode, FileAccessPer
         return ioStream.leakRef();
     }
 
-    if (mode == FileOpenMode::Read)
+    if (mode == FileOpenMode::Read || mode == FileOpenMode::ReadWrite)
         ioStream = adoptGRef(g_file_open_readwrite(file.get(), nullptr, nullptr));
     else {
-        ASSERT(mode == FileOpenMode::Truncate || mode == FileOpenMode::ReadWrite);
-        bool fileExists = g_file_test(filename.data(), static_cast<GFileTest>(G_FILE_TEST_EXISTS | G_FILE_TEST_IS_REGULAR));
-        if (fileExists) {
-            if (mode == FileOpenMode::ReadWrite)
-                ioStream = adoptGRef(g_file_open_readwrite(file.get(), nullptr, nullptr));
-            else {
-                deleteFile(path);
-                fileExists = false;
-            } 
-        }
-        if (!fileExists)
-            ioStream = adoptGRef(g_file_create_readwrite(file.get(), permissionFlag, nullptr, nullptr));
+        ASSERT(mode == FileOpenMode::Truncate);
+        ioStream = adoptGRef(g_file_replace_readwrite(file.get(), nullptr, FALSE, permissionFlag, nullptr, nullptr));
     }
 
     return ioStream.leakRef();
```

I'm building now to make sure that compiles.

(P.S. I also noticed that it's weird to convert FileOpenMode::Read to a read/write, but the return value has to be a GIOStream, which is inherently writable, so it's not actually possible to respect the FileOpenMode. To do actual read-only or write-only opens, we'd need to return GInputStream or GOutputStream instead. Then PlatformIOHandle would have to be a GObject rather than a GIOStream. Probably not worth it.)

mcatanzaro replied:
Fixed a build error. It doesn't immediately explode at runtime either, so decent chance it's even correct....

szewai replied:
Thanks for the review; I updated Source/WTF/wtf/glib/FileSystemGlib.cpp with the proposed approach.

szewai committed:
Rename FileSystem::FileOpenMode::Write to FileSystem::FileOpenMode::Truncate
https://bugs.webkit.org/show_bug.cgi?id=251073
rdar://104590894

Reviewed by Žan Doberšek.

In existing implementation, file opened with FileSystem::FileOpenMode::Write will be truncated or recreated if it
exists. This behavior is not clearly indicated in the name, and it causes confusion to users. For example, some places
check file size or truncate file after opening it with FileSystem::FileOpenMode::Write mode. They may actually need to
use FileSystem::FileOpenMode::ReadWrite mode. Therefore, this patch renames FileSystem::FileOpenMode::Write to
FileSystem::FileOpenMode::Truncate. Also, this patch makes glib implementation of FileSystem::FileOpenMode::Truncate
mode aligned with POSIX and win implementation, which recreates file when file with same name already exists.

* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/WasmStreamingParser.cpp:
(JSC::Wasm::dumpWasmSource):
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::MappedFileData::mapFileHandle):
(WTF::FileSystemImpl::readOrMakeSalt):
(WTF::FileSystemImpl::overwriteEntireFile):
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::FileSystemImpl::openFile):
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::openFile):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::openFile):
(WTF::FileSystemImpl::MappedFileData::mapFileHandle):
* Source/WebCore/Modules/webdatabase/OriginLock.cpp:
* Source/WebCore/loader/appcache/ApplicationCacheStorage.cpp:
(WebCore::ApplicationCacheStorage::writeDataToUniqueFileInDirectory):
* Source/WebCore/platform/network/BlobRegistryImpl.cpp:
(WebCore::BlobRegistryImpl::writeBlobToFilePath):
* Source/WebCore/platform/network/curl/CookieJarDB.cpp:
(WebCore::CookieJarDB::flagDatabaseCorruption):
* Source/WebCore/platform/network/curl/CurlCacheEntry.cpp:
(WebCore::CurlCacheEntry::saveResponseHeaders):
(WebCore::CurlCacheEntry::openContentFile):
* Source/WebCore/platform/network/curl/CurlCacheManager.cpp:
(WebCore::CurlCacheManager::saveIndex):
* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::store):
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::NetworkDataTaskBlob::download):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
(WebKit::ServiceWorkerDownloadTask::setPendingDownloadLocation):
* Source/WebKit/NetworkProcess/cache/CacheStorageEngine.cpp:
(WebKit::CacheStorage::Engine::writeSizeFile):
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::dumpContentsToFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCurl.cpp:
(WebKit::NetworkCache::IOChannel::IOChannel):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::invokeDidReceiveResponse):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::writeCachesList):
* Source/WebKit/Shared/PersistencyUtils.cpp:
(WebKit::writeToDisk):
* Source/WebKit/Shared/WebMemorySampler.cpp:
(WebKit::WebMemorySampler::initializeSandboxedLogFile):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::writeSandboxDataToCacheFile):
* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::ContentRuleListStore::invalidateContentRuleListVersion):
* Source/WebKit/UIProcess/ios/WKModelView.mm:
(-[WKModelView createFileForModel:]):
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.mm:
(WebKit::ARKitInlinePreviewModelPlayerMac::createFile):
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
(TestWebKitAPI::createTestFile):
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebCore/FileMonitor.cpp:
(TestWebKitAPI::TEST_F):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/DatabaseTrackerTest.mm:
(TestWebKitAPI::createFileAtPath):

Canonical link: https://commits.webkit.org/259689@main
